pygmt/tests/test_clib_to_numpy.py

```diff
@@ -0,0 +1,159 @@
+"""
+
+import numpy as np
+import numpy.testing as npt
+import pandas as pd
+import pytest
+from pygmt.clib.conversion import _to_numpy
+
+
+def _check_result(result, expected_dtype):
+    """
+    A helper function to check if the result of the _to_numpy function is a C-contiguous
+    NumPy array with the expected dtype.
+    """
+    assert isinstance(result, np.ndarray)
+    assert result.flags.c_contiguous
+    assert result.dtype.type == expected_dtype
+
+
+########################################################################################
+# Test the _to_numpy function with Python built-in types.
+########################################################################################
+@pytest.mark.parametrize(
+    ("data", "expected_dtype"),
+    [
+        pytest.param([1, 2, 3], np.int64, id="int"),
+        pytest.param([1.0, 2.0, 3.0], np.float64, id="float"),
+    ],
+)
+def test_to_numpy_python_types_numeric(data, expected_dtype):
+    """
+    Test the _to_numpy function with Python built-in numeric types.
+    """
+    result = _to_numpy(data)
+    _check_result(result, expected_dtype)
+    npt.assert_array_equal(result, data)
+
+
+########################################################################################
+# Test the _to_numpy function with NumPy arrays.
+#
+# There are 24 fundamental dtypes in NumPy. Not all of them are supported by PyGMT.
+#
+# - Numeric dtypes:
+#   - int8, int16, int32, int64, longlong
+#   - uint8, uint16, uint32, uint64, ulonglong
+#   - float16, float32, float64, longdouble
+#   - complex64, complex128, clongdouble
+# - bool
+# - datetime64, timedelta64
+# - str_
+# - bytes_
+# - object_
+# - void
+#
+# Reference: https://numpy.org/doc/2.1/reference/arrays.scalars.html
+########################################################################################
+@pytest.mark.parametrize(
+    ("dtype", "expected_dtype"),
+    [
+        pytest.param(np.int8, np.int8, id="int8"),
+        pytest.param(np.int16, np.int16, id="int16"),
+        pytest.param(np.int32, np.int32, id="int32"),
+        pytest.param(np.int64, np.int64, id="int64"),
+        pytest.param(np.longlong, np.longlong, id="longlong"),
+        pytest.param(np.uint8, np.uint8, id="uint8"),
+        pytest.param(np.uint16, np.uint16, id="uint16"),
+        pytest.param(np.uint32, np.uint32, id="uint32"),
+        pytest.param(np.uint64, np.uint64, id="uint64"),
+        pytest.param(np.ulonglong, np.ulonglong, id="ulonglong"),
+        pytest.param(np.float16, np.float16, id="float16"),
+        pytest.param(np.float32, np.float32, id="float32"),
+        pytest.param(np.float64, np.float64, id="float64"),
+        pytest.param(np.longdouble, np.longdouble, id="longdouble"),
+        pytest.param(np.complex64, np.complex64, id="complex64"),
+        pytest.param(np.complex128, np.complex128, id="complex128"),
+        pytest.param(np.clongdouble, np.clongdouble, id="clongdouble"),
+    ],
+)
```

Comment:
> This test is almost the same as `test_to_numpy_ndarray_numpy_dtypes_numeric` and can be combined into one single test, but I feel two separate tests are more readable and maintainable.

Can reduce duplication by creating a variable with the list of parametrized options, see https://docs.pytest.org/en/6.2.x/example/parametrize.html#paramexamples


```suggestion
np_dtype_params = [
    pytest.param(np.int8, np.int8, id="int8"),
    pytest.param(np.int16, np.int16, id="int16"),
    pytest.param(np.int32, np.int32, id="int32"),
    pytest.param(np.int64, np.int64, id="int64"),
    pytest.param(np.longlong, np.longlong, id="longlong"),
    pytest.param(np.uint8, np.uint8, id="uint8"),
    pytest.param(np.uint16, np.uint16, id="uint16"),
    pytest.param(np.uint32, np.uint32, id="uint32"),
    pytest.param(np.uint64, np.uint64, id="uint64"),
    pytest.param(np.ulonglong, np.ulonglong, id="ulonglong"),
    pytest.param(np.float16, np.float16, id="float16"),
    pytest.param(np.float32, np.float32, id="float32"),
    pytest.param(np.float64, np.float64, id="float64"),
    pytest.param(np.longdouble, np.longdouble, id="longdouble"),
    pytest.param(np.complex64, np.complex64, id="complex64"),
    pytest.param(np.complex128, np.complex128, id="complex128"),
    pytest.param(np.clongdouble, np.clongdouble, id="clongdouble"),
]


@pytest.mark.parametrize(("dtype", "expected_dtype"), np_dtype_params)
```